service/src/LevelZero.cpp

```diff
@@ -523,6 +526,366 @@ namespace geopm
             }
         }
     }
+    void LevelZeroImp::metric_group_cache(unsigned int device_idx) {
```

Comment:
On a separate note about this method, I feel it's a bit lengthy and requires drilling down into nested blocks, so it's not straightforward to quickly grok what it's doing, so if you can, please break it down a bit.